hazelcast/src/hazelcast/client/spi.cpp

```diff
@@ -1321,13 +1360,44 @@ namespace hazelcast {
                     if (!msg) {
                         BOOST_THROW_EXCEPTION(exception::IllegalArgumentException("response can't be null"));
                     }
+
+                    int8_t expected_backups = msg->get_number_of_backups();
+
+                    // if a regular response comes and there are backups, we need to wait for the backups
+                    // when the backups complete, the response will be send by the last backup or backup-timeout-handle mechanism kicks on
+                    if (expected_backups > backup_acks_received_) {
+                        // so the invocation has backups and since not all backups have completed, we need to wait
+                        // (it could be that backups arrive earlier than the response)
+
+                        pending_response_received_time_ = std::chrono::steady_clock::now();
+
+                        backup_acks_expected_ = expected_backups;
+
+                        // it is very important that the response is set after the backupsAcksExpected is set, else the system
+                        // can assume the invocation is complete because there is a response and no backups need to respond
+                        pending_response_ = msg;
+
+                        if (backup_acks_received_ != expected_backups) {
```

Comment:
Is this if statement necessary? We went into this block because `expected_backups > backup_acks_received_` was true, so `backup_acks_received_ != expected_backups` should also be true. 
I assume there isn't any other thread modifying these variables as we move from line 1368 to here, since it seems all things related to this invocation run within the same strand. 
I also checked the Java impl but these variables was declared volatile there, meaning that there is some concurrency at play, but its not true for our case.

Reply:
Yes, that seems an unneeded check. it can also be deleted at Java client. I will put a return statement there without the check.

As for non-atomic variables, yes, I do this on purpose, since I always access these variables from the same io thread (socket strand), hence i do not need atomic variables. If any access is needed I always do posting through the same io thread.